src/nccl_ofi_net.c

```diff
@@ -188,34 +189,95 @@ int nccl_net_ofi_create_plugin(nccl_net_ofi_plugin_t **plugin_p)
 			goto exit;
 	}
 
-	/* Select and initialize protocol data structure.
-	 * platform_init() may change the default, so this must occur
-	 * after the platform init call.
+	/* This is ugly, but here's the basic protocol selection
+	 * logic:
+	 *   1. if the user set NCCL_OFI_PROTOCOL, use that.
+	 *   2. if the platform init set nccl_ofi_selected_protocol,
+	 *      use that.
+	 *   3. If the rdma protocol reports multiple nics per device
+	 *      and initialized successfully, use that.
+	 *   4. If the sendrecv protocol initialized successfully, use
+	 *      that
+	 *   5. If the rdma protocol initialized successfully, use
+	 *      that.
 	 */
 	if (ofi_nccl_protocol()) {
 		nccl_ofi_selected_protocol = ofi_nccl_protocol();
 		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s (user set)",
 			      nccl_ofi_selected_protocol);
-	} else {
-		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s",
+	} else if (nccl_ofi_selected_protocol != NULL) {
+		NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Using transport protocol %s (platform set)",
 			      nccl_ofi_selected_protocol);
 	}
 
-	if (0 == strcasecmp(nccl_ofi_selected_protocol, "SENDRECV")) {
-		ret = nccl_net_ofi_sendrecv_init(provider_filter, plugin_p);
-		if (ret != 0) {
-			NCCL_OFI_WARN("Failed to initialize sendrecv protocol");
+	if (nccl_ofi_selected_protocol != NULL) {
+		bool dummy;
+
+		if (0 == strcasecmp(nccl_ofi_selected_protocol, "SENDRECV")) {
+			ret = nccl_net_ofi_sendrecv_init(provider_filter, &plugin);
+			if (ret != 0) {
+				NCCL_OFI_WARN("Failed to initialize sendrecv protocol");
+				goto exit;
+			}
+		} else if (0 == strcasecmp(nccl_ofi_selected_protocol, "RDMA")) {
+			ret = nccl_net_ofi_rdma_init(provider_filter, &plugin, &dummy);
+			if (ret != 0) {
+				NCCL_OFI_WARN("Failed to initialize rdma protocol");
+				goto exit;
+			}
+		} else {
+			NCCL_OFI_WARN("Unable to find plugin protocol %s", nccl_ofi_selected_protocol);
+			ret = -ENOTSUP;
 			goto exit;
 		}
-	} else if (0 == strcasecmp(nccl_ofi_selected_protocol, "RDMA")) {
-		ret = nccl_net_ofi_rdma_init(provider_filter, plugin_p);
+	} else {
+		bool have_multiple_rails = false;
+		nccl_net_ofi_plugin_t *rdma_plugin = NULL, *sendrecv_plugin = NULL;
+
+		ret = nccl_net_ofi_rdma_init(provider_filter, &rdma_plugin, &have_multiple_rails);
 		if (ret != 0) {
-			NCCL_OFI_WARN("Failed to initialize rdma protocol");
+			NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+				       "Failed to initialize rdma protocol: %d", fi_strerror(-ret));
```

Comment:
Probably copied from somewhere, but this should be `%s`, not `%d`?

I wish we would get a compiler warning for this.

Reply:
Also, is `fi_strerror` really the right thing to call here? Since we return general errno codes, maybe just `strerror` makes more sense here?

Reply:
https://github.com/aws/aws-ofi-nccl/pull/615